### PR TITLE
remove non-dependency related template functions

### PR DESCRIPTION
**Pull request by eikenb**
Remove all the utility/helper template functions that were in
consul-template but aren't the focus of hashicat. With hashicat we want
to focus on the external dependencies/data, monitoring and triggering
re-renders. Keeping the helper functions needed to extend that for a
specific application to that application.

This strips out all the functions and their tests.

*REVIEWERS* .. please chime in with your thoughts on the dependency oriented template functions (byMeta, byKey, byTag) and about if it is useful to keep Environment variables as an "external" source (ie. that eng should be kept in hashicat). Thanks!

Fixes: #15

### CODE REVIEW

**Comment by eikenb** [comment]
To keep `byMeta` and friends in consul-template, this will require exporting the internal data types used so that they can be used in the consul-template tests for these functions.

**Comment by eikenb** [comment]
Note we (kim and I) had talked about how the types of the internal dependencies are going to need to be public to allow certain types of custom template functions to be implemented... basically sync has the exact same issue here.

IMO this is a strong argument that they shouldn't be in hashicat, that they should stay in consul-template as keeping them there surfaces this issue that might have been elided over (had sync not run into it as well).

**Comment by lornasong** [comment]
@eikenb - I haven't really formed any opinions but wanted to make sure I understand:

>  types of the internal dependencies are going to need to be public to allow certain types of custom template functions to be implemented

Is this the case even if you include `byMeta` and friends in hashicat or this is the cases that arises by not including `byMeta` and friend in hashicat? Thanks!

**Comment by eikenb** [comment]
@lornasong  said..
> Is this the case even if you include `byMeta` and friends in hashicat or this is the cases that arises by not including `byMeta` and friend in hashicat? Thanks!

If we include `byMeta` (et al.) in hashicat, then they can reference the internal data structures, and so you wouldn't need to make the data structures they use public. *But* the problem is that others need to be able to access those structures for the same reasons `byMeta` does and we didn't realize that until Kim came across the problem in Sync (I would have as well in consul-template, but hadn't got to that yet).

So whether we include `byMeta` (et al.) in hashicat or not we still will need to make those data structures public. I liked how *not* including `byMeta` we other cases to make sure that works correctly. I like having as many as possible use cases for each feature so I can have more confidence in the generality of it.

**Comment by findkim** [comment]
Hm this is pretty difficult decision for me on trimming down hcat from all the non-essential template functions related to dep objects. I do appreciate the helper functions that aren't tied to dep/internal objects as a standard library option. I can imagine other projects quickly copy/pasta several of these utility functions or end up having a non-consistent API across projects for templates because it the same function name but different specs/params. If these are used as a base API, and other projects add to it with external functions specific to its use-case, there would be a large foundation of template consistency from these utility functions.

As for your specific question on `byMeta` (et al.): keeping these dep/internal specific object functions in hcat is useful because it co-locates template functions that are strongly tied to those object structures.

**Comment by lornasong** [comment]
Thanks for taking the time to explain this @eikenb! Since including `byMeta` + friends still requires making those data structures public makes me lean towards removing them or having them separate somewhere else. I don’t know all the considerations though. Will be lurking to follow along to follow the discussion :)

**Comment by eikenb** [comment]
> I do appreciate the helper functions that aren't tied to dep/internal objects as a standard library option.

Maybe a new sub-module that would consist solely of template functions that can be optionally added, either individually or in groups. Splitting them out in a sub-module would keep them handy to applications while having them be separate, so they wouldn't have access to private fields and would require the same changes as if the clients were supplying them. But they'd still be included as the standard implementation.

This would allow the smooth workflow of upstreaming the best custom template functions that have been developed for a specific application. It could be upstreamed without change to the application (until they wanted) as the custom local one could still have precedence meaning we wouldn't have to worry about collisions.

If we do this would we want all the current functions or are there any that look particularly consul-templat-y and that we'd want to leave there?

**Comment by findkim** [comment]
Sub-module is an interesting option! Would be a good compromise to keep it independent from the crux of hcat logic.

The [helpers](https://github.com/hashicorp/consul-template#helper-functions) seem like a good set for one sub-module (excluding the ones you have selected for this PR). The ones that strike me as consul-template-y are the [`scratch`](https://github.com/hashicorp/consul-template#scratch) set. Projects using the library would effectively have run-time process for temp storage and calculations, where as template writers for CT need this as a stop-gap.

**Comment by eikenb** [comment]
Quick slack chat later and we decided to go forward with most templates in a new sub-module. They will all exist at one level, organized with files with each function available to include separately or in groups for including batches.

Everything except the `scratch` functions will be included in these in hashicat. Though while porting I'll review them as I go to see if any others don't make sense.

I'm going to continue with this PR and open a new issue/PR for adding the functions back as a library.

File: template.go [hunk 1]
```diff
@@ -210,60 +208,12 @@ func funcMap(i *funcMapInput) template.FuncMap {
...
-		"sockaddr":        sockaddr,
-		// Math functions
-		"add":      add,
-		"subtract": subtract,
-		"multiply": multiply,
-		"divide":   divide,
-		"modulo":   modulo,
-		"minimum":  minimum,
-		"maximum":  maximum,
+		// ???
```
Comment (findkim) [inline template.go:211]: > ???

Very interested in its final form

**Review by findkim — APPROVED** [review]
Thanks for opening up the discussion on these template functions! I like the direction hcat is going with submodules and as always, we can revisit a function or promote other functions to make it canon in the future.

File: template.go (hunk 1 above)
Comment (eikenb) [reply template.go:211]: Right. Got thinking about the library idea and forgot that stuff was still in there. I'll strip it out.

Comment (eikenb) [reply template.go:211]: A bit more to it with tests and trying to move things into tfunc/ instead of just deleting them. I'll have this done in a bit.

Comment (eikenb) [reply template.go]: These are not stripped out. It should be merge-ready now.

Comment (findkim) [reply template.go]: The latest changes I can see only include deletes -- are you still planning to move them to `tfunc`? if so might be missing those files 😄

Comment (eikenb) [reply template.go]: not/now ... "now stripped out"

**Comment by eikenb** [comment]
With the standard template library PR merged I think this is good to go. Note that once merged we'll need to explicitly include the template functions in CTS.

**Pull request merged**